### PR TITLE
Upgrade Demo

### PR DESCRIPTION
**Pull request by vgoliber**
Heading out on vacation so wanted to give you all a chance to review.

*Note: I know that it needs more unit tests, but wanted to get this through with some basic ones for now. I can continue to add more after.

### CODE REVIEW

File: app.py
Comment (JoelPasvolsky) [inline app.py:342]: This Employee Schedule screen needs additional legend items: X and P (are only described in the previous Availability screen) and "-" (dash) is not described anywhere

File: app.py
Comment (JoelPasvolsky) [inline app.py:124]: As I say in another comment, I think it would look nicer to use a tab to separate basic and additional configuration options. If you decide to stick with the collapsable button, I think it should at least use a different color to the "Solve CQM" button, which should stand out.  Currently the two buttons are identical:

image

File: app.py
Comment (JoelPasvolsky) [inline app.py:524]: Currently, changing inputs such as that the Employee Availability schedule changes does not blank out an existing Employee Schedule. So one gets a mismatch:

 
image
image

Be nice to blank out the existing result when the user changes the inputs.

**Review by JoelPasvolsky — COMMENTED** [review]
@vgoliber, this is a really nice upgrade of the example's user interface. 

Here are my comments from a first look, so these should provide my main feedback. I will try to spend more time on a low-level code review for things like unit testing etc. but wanted to get back to you quickly here with the most high-impact items. I'll first go over the readme file.

File: README.md
Comment (JoelPasvolsky) [inline README.md:25]:
```suggestion
### Inputs
Input options under the Basic tab are as follows:

- Number of employees: Schedules always include 2 managers  and 1 trainee.
- Random seed: Optional; use if you want consistency between subsequent runs of the example.
- Example scenario: Auto-populates all settings with scenarios of varying
  sizes that produce feasible solutions.

Additional input options under the More Options tab are as follows:

- Allow isolated days off: Unchecked, this option means that employees are
  scheduled at least two consecutive days off between work days. 
- Require a manager on every shift: Checked, this option means that every shift
  must have exactly one manager on duty to supervise.
- Min/max shifts per employee: The range you set determines the number of shifts an
  employee can work in the month.
- Min/max employees per shift: The range you set determines how many employees need
  to be assigned to each shift.
- Max consecutive shifts: The maximum number of consecutive shifts an employee
  can be scheduled before a day off must be scheduled.

### Outputs

Once the problem has completed, the best solution returned is displayed in
place of "Employee Availability". Click back to "Employee Availability" using
the tabs at the top of the schedule card.
```

Recently Murray updated our [example template](https://github.com/dwave-examples/template) and Catherine has begun updating examples to fit the template, so it's worth adhering to the template with new/refreshed examples

Note that you can edit this suggested update (before or after accepting) if you do not accept the earlier suggestions on design using tabs instead of the collapsable button or want different text.

File: README.md
Comment (JoelPasvolsky) [inline README.md:12]:
```suggestion
a scheduling scenario with a variety of employees and rules.

![Screen Image](assets/demo_image.png)
```

In keeping with our recently updated [example template](https://github.com/dwave-examples/template), this is the recommended position for your screenshot.

File: README.md
Comment (JoelPasvolsky) [inline README.md:30]:
```suggestion
The solution returned is either the best feasible solution (if a feasible
solution is found) or the best infeasible solution (if no feasible solution is
found). If an infeasible solution is found, scrolling down shows the list of
constraints that were violated.
```

File: README.md
Comment (JoelPasvolsky) [inline README.md:35]:
```suggestion
```

In keeping with our recently updated [example template](https://github.com/dwave-examples/template), the recommended position for your screenshot is at the top and then this text is not needed.

File: README.md
Comment (JoelPasvolsky) [inline README.md:37]:
```suggestion
```

In keeping with our recently updated [example template](https://github.com/dwave-examples/template), the recommended position for your screenshot is at the top.

File: README.md
Comment (JoelPasvolsky) [inline README.md:39]:
```suggestion
```

The following text will be under the template's ``Inputs`` header three (if you go with the previous suggested changes)

File: README.md
Comment (JoelPasvolsky) [inline README.md:43]:
```suggestion
unavailable shifts are in orange and marked with an 'X'.
```

I first assumed that "organe" was some color name I was not familiar with :-)

File: README.md
Comment (JoelPasvolsky) [inline README.md:41]: This should also go under the Inputs header but I'll leave that to you

File: README.md
Comment (JoelPasvolsky) [inline README.md:65]:
```suggestion
```

Should be moved to under ###Inputs per the updated [example template](https://github.com/dwave-examples/template)

I edited the phrasing in a separate suggested change under ###Inputs

File: README.md
Comment (JoelPasvolsky) [inline README.md:68]:
```suggestion
```

See suggested change under ###Inputs

File: README.md
Comment (JoelPasvolsky) [inline README.md:80]:
```suggestion
```

Should be moved to under ###Inputs per the updated [example template](https://github.com/dwave-examples/template)

I edited the phrasing in a separate suggested change under ###Inputs

**Review by JoelPasvolsky — COMMENTED** [review]
@vgoliber, in this review of the README I made a few suggested changes and some comments mostly intended to keep it compliant to our [example template](https://github.com/dwave-examples/template) that Murray recently updated. 

You probably know this: you can collect all suggested changes you wish to accept into a single commit. You might have the access to directly edit suggested changes if they are not quite what you want and one can always accept & commit and then edit the updated text. I expect that if you accept the suggested changes, there will be a few minor edits needed after to get the flow just right again

File: app.py
Comment (vgoliber) [reply app.py:580]: I could use some help here - I tried to replicate what you did in your tour planning demo but it got complex pretty quick and I abandoned that approach. Is this something that you could help me with?

Comment (JoelPasvolsky) [reply app.py:580]: Sure. I remember spending several days to get the submission with its time and colors working but what I don't remember is how modular it might or might not be. A much quicker and simpler alternative to start with can be adding a simple text window under the Solve CQM button to print out the submission messages. 

What is your schedule for working on this like? Would you like me to clone the example and make a PR to your branch at some point to add such an indication? Let me know what works best for you and when I can be of help.

Comment (vgoliber) [reply app.py:580]: I'm out most of tomorrow and Monday, but plan to work through the rest of these on Tuesday. I think the goal was to get these done before 23 January. If you can do the PR before Tuesday that would be fantastic, but I know that's a short timeline for you.

Comment (JoelPasvolsky) [reply app.py:580]: Okay, I can try getting you something to start with by EOD Monday so it won't conflict with you making changes. Once your updates are mostly done, we can improve it as needed.

Comment (JoelPasvolsky) [reply app.py:580]: Now that this update of the example is not scheduled for the week after next, I've been pulled off to other work so not fully tested (though seems about right) but here are two PRs:

* [PR that adds indicator and uses tabs for advanced options](https://github.com/vgoliber/employee-scheduling-1/pull/2). I recommend merging just this one and then updating to your liking
* [PR that adds just the indicator](https://github.com/vgoliber/employee-scheduling-1/pull/1). I recommend declining in favor of the other PR

**Comment by vgoliber** [comment]
I believe I have addressed all of the feedback in this review. Let me know if any other edits are needed before merging.

File: requirements.txt
Comment (randomir) [inline requirements.txt:5]: Perhaps loosen up requirements on packages that come preinstalled in the Ocean dev container, like ocean and pandas.

**Comment by arcondello** [comment]
Added @thisac , mostly for visibility

**Review by thisac — COMMENTED** [review]
Thanks @vgoliber! Nice to get a Dash front for this demo. 🎉 

I've run it but haven't tested it very rigorously. A few general comments (many of them a bit nit-picky; I understand that you might still be updating it).

* ~The text of the tabs isn't visible due to white-on-white colour.~ Fixed by updating `dash-bootstrap-components`.
* Both background and table jump around when switching between tabs. I think some CSS magic might be necessary to make things look nice with different layouts, window sizes, etc. Let me know if you need some help with this.
* Could be nice to add the `Mgr` and `Tr` descriptions to a legend (minor).
* Table doesn't seem to update when using the arrows to change values for e.g., number of employees and seed (works when entering a number or clicking the box).
* There's a resize icon (with which you can resize the upper div, although it's note useful) which should probably be hidden.
image
* The background image should probably be higher resolution.

File: employee_scheduling.py
Comment (randomir) [inline employee_scheduling.py:22]: In fact, if `availability` is a `dict` (as it seems to be):
```suggestion
    employees = list(availability)
```

**Comment by vgoliber** [comment]
@thisac: in response to this comment "Table doesn't seem to update when using the arrows to change values for e.g., number of employees and seed (works when entering a number or clicking the box)."

This is a known issue with dbc.Input and doesn't have an obvious fix that I can see.

See: https://github.com/plotly/dash-core-components/issues/809

**Review by JoelPasvolsky — APPROVED** [review]
LGTM. I don't have time for a second full review anytime soon, I checked the updated README and did a few quick runs. A few not-insignificant bugs popped out, @vgoliber & @thisac can discuss whether to fix before or after merging:

1. If you switch the open tab from Availability to Schedule during the CQM submission (after clicking Submit but before results are returned), the resulting schedule is not displayed
2. The Availability table is updated if you set the number of employees and hit Enter but if you click to increment/decrement (or do not hit Enter), the table remains outdated
3. Not all infeasible schedules are caught; for example, if you set the min/max shifts to 31, the resulting schedule is feasible even though employees have <31 shifts because of days they are not available

**Comment by vgoliber** [comment]
@JoelPasvolsky - Theodor had the same question as your item 2. See my response: https://github.com/dwave-examples/employee-scheduling/pull/18#issuecomment-1978471790

**Comment by vgoliber** [comment]
Thanks for finding those @JoelPasvolsky!

1. I think I fixed this but please check and confirm.
2. By switching to sliders I've think I've worked around the unexpected behavior pointed out by both you and @thisac. Please take a look and see if this format works ok as a replacement.
3. Fixed this issue - good catch!

**Pull request merged**
